### PR TITLE
Add dune buildsystem

### PR DESCRIPTION
**Pull request by andlaus**
This PR is related to #277, but it does not remove the current build system. Instead, it just adds everything required to use DUNE's build system and the `USE_OPM_BUILD_SYSTEM` option to select which of those two shall be used. The default is to stick with the current system. (except when running under dunecontrol, because in that case you most probably *really* want to use DUNE's system.)

The reasons why the dune system might want to be used are the same as for #277:

- Playing the usual whack-a-mole game for new Dune releases is not necessary anymore
- It makes it significantly easier to play around with less common libraries that are not (yet?) relevant for `flow` but already have support in Dune build system but not by the OPM one. Examples are dune-fem, dune-spgrid, AlbertaGrid, PTScotch, METIS, VC, UG and quite a few more.
- The dune build system is more widely tested, e.g., it should work fine on most supercomputers or with some alternative compilers like `icc`.

### CODE REVIEW

**Comment by atgeirr** [comment]
Is this the PR that was discussed earlier today in the hangout?

**Comment by andlaus** [comment]
yes. may I point you to the fact, that it only adds about 600 lines and modifies just 16. I guess it is hard to get a build system with all bells and whistles in a less obstructive way.

**Comment by atgeirr** [comment]
> yes. may I point you to the fact, that it only adds about 600 lines and modifies just 16

The way it was discussed, I had the impression that it would be only a handful of lines something like 6 rather than 600...!

**Comment by andlaus** [comment]
600 is only "a hand full" compared to the 40000 of the current build system, don't you agree?

also, most of these 600 lines are spend to provide a slightly modified version of the `opm_add_test()` macro which is currently provided by `OpmSatellites.cmake` and thus cannot be used in the dune build system context.

**Comment by atgeirr** [comment]
> 600 is only "a hand full" compared to the 40000 of the current build system, don't you agree?

No. `wc cmake/Modules/*.cmake` yields 7672 lines of code (yes it is still a lot, and there are other parts), but 600 is certainly not just a minor change!

**Comment by dr-robertk** [comment]
@andlaus  and @atgeirr : The discussion about how many lines are used is a bit childish. 

Within the activities of the IOR Centre and also for some of the latest things on performance of flow this option to use the DUNE build system has been most helpful as we want to bring in some already existing functionality from DUNE. With the current OPM build system this is not possible, and before somebody claims otherwise, I have tried for more than 2 years now. The problems vary from dependencies to shared  libraries added (that were never there) to module dependencies being wrong and version number not being detected etc. That is just some of the problems, not to mention the problematic sibling build which continuously fails to recognize the correct installation of libecl and so on. 

The option to use the DUNE build system should be seen as an option, not more. I can't see anything wrong with that and it would be very sad if there was no room for such experimental extensions inside the OPM framework. 

The changes introduced in these PRs should not touch or change the OPM build system at all, otherwise it can be seen as a mistake and in that case it will be fixed. In addition, both build systems share list of source files and executables listed in CMakeList_files.cmake in each module such that files lists only have to be edited in one place.

**Comment by andlaus** [comment]
> No. wc cmake/Modules/*.cmake yields 7672 lines of code (yes it is still a lot, and there are other parts), but 600 is certainly not just a minor change!

yes, I confused this with the total number of lines of code for some module. I still get more almost 10000 for the current BS in opm-common alone, though:

```
    15 ./cmake/Scripts/RemoveEmptyDir.cmake
    65 ./cmake/Scripts/WriteVerSHA.cmake
    25 ./cmake/Scripts/DuneCompat2.cmake
    99 ./cmake/Scripts/CheckCommits.cmake
   739 ./cmake/OPM-CMake.md
   169 ./cmake/Modules/LibtoolArchives.cmake
    51 ./cmake/Modules/Finddune-geometry.cmake
    49 ./cmake/Modules/FindQuadmath.cmake
    42 ./cmake/Modules/Findopm-grid.cmake
    25 ./cmake/Modules/FindC99.cmake
   287 ./cmake/Modules/FindERT.cmake
    83 ./cmake/Modules/Finddune-common.cmake
   103 ./cmake/Modules/OpmStaticTargets.cmake
   307 ./cmake/Modules/compat-2.8.7/FindLAPACK.cmake
   624 ./cmake/Modules/compat-2.8.7/FindBLAS.cmake
    61 ./cmake/Modules/compat-2.8.7/CMakePushCheckState.cmake
    47 ./cmake/Modules/compat-2.8.5/FindGit.cmake
    97 ./cmake/Modules/OpmCompile.cmake
    64 ./cmake/Modules/Finddune-grid.cmake
    91 ./cmake/Modules/OpmDoc.cmake
    50 ./cmake/Modules/UseOnlyNeeded.cmake
    23 ./cmake/Modules/FindCxaDemangle.cmake
    49 ./cmake/Modules/FindMETIS.cmake
    36 ./cmake/Modules/UseDynamicBoost.cmake
   106 ./cmake/Modules/OpmFiles.cmake
    24 ./cmake/Modules/Findopm-data.cmake
    44 ./cmake/Modules/Findopm-output.cmake
   138 ./cmake/Modules/compat-2.8.3/CMakeParseArguments.cmake
    86 ./cmake/Modules/Finddune-alugrid.cmake
    24 ./cmake/Modules/opm-output-prereqs.cmake
    11 ./cmake/Modules/FindCppCheck.cmake
   102 ./cmake/Modules/AddOptions.cmake
    44 ./cmake/Modules/Findopm-simulators.cmake
    50 ./cmake/Modules/opm-core-prereqs.cmake
    43 ./cmake/Modules/Findopm-material.cmake
   132 ./cmake/Modules/UseDebugSymbols.cmake
    90 ./cmake/Modules/Findcjson.cmake
    53 ./cmake/Modules/UseThreads.cmake
   102 ./cmake/Modules/UseVCSInfo.cmake
    81 ./cmake/Modules/OpmGrid.cmake
   154 ./cmake/Modules/OpmProject.cmake
    48 ./cmake/Modules/Findopm-flowdiagnostics.cmake
    42 ./cmake/Modules/Findewoms.cmake
    58 ./cmake/Modules/FindZOLTAN.cmake
    11 ./cmake/Modules/UseFastBuilds.cmake
   179 ./cmake/Modules/FindSuperLU.cmake
    32 ./cmake/Modules/opm-upscaling-prereqs.cmake
    73 ./cmake/Modules/OpmInit.cmake
    80 ./cmake/Modules/FindPTScotch.cmake
   198 ./cmake/Modules/UseSystemInfo.cmake
    66 ./cmake/Modules/Findopm-parser.cmake
    32 ./cmake/Modules/UseDuneVer.cmake
    49 ./cmake/Modules/FindCwrap.cmake
   112 ./cmake/Modules/FindUG.cmake
    17 ./cmake/Modules/opm-common-prereqs.cmake
    43 ./cmake/Modules/Findopm-common.cmake
    50 ./cmake/Modules/UseStaticAnalysis.cmake
   166 ./cmake/Modules/FindPETSc.cmake
    42 ./cmake/Modules/Finddune-localfunctions.cmake
    41 ./cmake/Modules/Findopm-upscaling.cmake
   110 ./cmake/Modules/ConfigVars.cmake
    25 ./cmake/Modules/FindValgrind.cmake
    11 ./cmake/Modules/FindClangCheck.cmake
    20 ./cmake/Modules/opm-verteq-prereqs.cmake
   294 ./cmake/Modules/FindSuiteSparse.cmake
    62 ./cmake/Modules/Finddune-istl.cmake
   125 ./cmake/Modules/FindEigen3.cmake
   205 ./cmake/Modules/OpmPackage.cmake
    96 ./cmake/Modules/FindParMETIS.cmake
    43 ./cmake/Modules/ewoms-prereqs.cmake
   364 ./cmake/Modules/OpmSatellites.cmake
   108 ./cmake/Modules/UseCompVer.cmake
   104 ./cmake/Modules/UseVersion.cmake
    39 ./cmake/Modules/opm-simulators-prereqs.cmake
    89 ./cmake/Modules/OpmDistClean.cmake
    99 ./cmake/Modules/OpmDefaults.cmake
    53 ./cmake/Modules/FindERTPython.cmake
    58 ./cmake/Modules/OpmKnown.cmake
    19 ./cmake/Modules/UseWarnings.cmake
    38 ./cmake/Modules/Duplicates.cmake
    36 ./cmake/Modules/opm-grid-prereqs.cmake
    57 ./cmake/Modules/Finddune-fem.cmake
    21 ./cmake/Modules/opm-flowdiagnostics-applications-prereqs.cmake
    18 ./cmake/Modules/opm-flowdiagnostics-prereqs.cmake
   242 ./cmake/Modules/OpmFind.cmake
    25 ./cmake/Modules/opm-material-prereqs.cmake
    47 ./cmake/Modules/DuneCompat.cmake
    48 ./cmake/Modules/Findopm-core.cmake
    35 ./cmake/Modules/UseMultiArch.cmake
    36 ./cmake/Modules/OpmSiblingSearch.cmake
   171 ./cmake/Modules/UsePrecompHeaders.cmake
    70 ./cmake/Modules/UseOpenMP.cmake
    51 ./cmake/Modules/Findlibrt.cmake
   468 ./cmake/Modules/FindCXX11Features.cmake
    90 ./cmake/Modules/UseFortranWrappers.cmake
    68 ./cmake/Modules/UseOptimization.cmake
   317 ./cmake/Modules/OpmLibMain.cmake
    27 ./cmake/Modules/OpmAliases.cmake
     7 ./cmake/Modules/OpmPythonTest.cmake
    64 ./cmake/Modules/OpmInstall.cmake
    41 ./cmake/Modules/Findopm-verteq.cmake
    28 ./cmake/Templates/opm-project-config-version.cmake.in
    73 ./cmake/Templates/opm-project-config.cmake.in
    86 ./CMakeLists_files.cmake
    54 ./CMakeLists.txt
  9966 total
```

that's still factor 15! (and please still keep in mind that of those 600 added lines about 350 are just stuff ported over from the current OPM build system.

be aware that in any case I will keep updating these branches even if they are not merged, i.e regardless of the outcome of this story, I will *not* waste any more of my time to trouble shoot the current BS. (I'm perfectly fine if the reverse applies for other people.)

to summarize: I'm of the strong opinion that it is always a good idea to go with the technically best solution. maybe this best solution is the current system, but blocking any attempts to find this out on political/emotional grounds is not very smart in my unhumble opinion. (To see my point, I'd also ask anyone who still uses `flow_legacy` by default to raise their hand. IMO, these people can argue to not merge these PRs with a straight face.)

**Comment by atgeirr** [comment]
I don't think it is childish to state the simple fact that this is not a minor change. Nor to provide actual linecounts, when linecounts are used as arguments. It came about because the change was presented as being only a few lines, and then turned out to be much more.

> be aware that in any case I will keep updating these branches even if they are not merged, i.e regardless of the outcome of this story, I will not waste any more of my time to trouble shoot the current BS. (I'm perfectly fine if the reverse applies for other people.)

Sounds like a reasonable way forward, doesn't it? If you don't merge, I won't close...

**Comment by andlaus** [comment]
> Sounds like a reasonable way forward, doesn't it? If you don't merge, I won't close...

agreed. Merging this would save everyone who needs the dune option some time though, because these branches do not need to be constantly rebased on top of master anymore.

**Comment by andlaus** [comment]
(before anyone asks: no, changes in master do not cause any merge conflicts w.r.t. the dune build system, but rebasing is required to get them.)

**Comment by dr-robertk** [comment]
@atgeirr I didn't particularly mean your comment about the lines but since you bought Andreas' bait you were definitely in on the line business.

**Comment by alfbr** [comment]
> Examples are dune-fem, dune-spgrid, AlbertaGrid, PTScotch, METIS, VC, UG and quite a few more

Actually, I believe PTScotch and METIS is already there. If you miss one, chances are it is not in Dune either. Are there particular ones you would like to add from Dune, all of them maybe? Should we make it easier to add new dependencies? The problem have been maintaining the new dependencies, and I am not thinking of the build system. Things like SuperLU have caused a lot of pain because things break between versions, and it has no proper packaging. As long as the dependencies are optional, and not linked by default, it is less of a problem. As for dune-fem and dune-spgrid, I thought those were almost trivial to add in OPM today whether you use dunecontrol or cmake directly. What exactly was the problem there?

**Comment by andlaus** [comment]
> Actually, I believe PTScotch and METIS is already there.

maybe, I haven't checked lately.

> Are there particular ones you would like to add from Dune

IIRC there recently was an issue with SuperLU on dune 2.5. I don't know if it was fixed, though.

> Things like SuperLU have caused a lot of pain because things break between versions, and it has no proper packaging.

I agree. that's why we should avoid the what-a-mole game with dune.

> As for dune-fem and dune-spgrid, I thought those were almost trivial to add in OPM today whether you use dunecontrol or cmake directly. What exactly was the problem there?

the problem is that it does not work out of the box. what about e.g. dune-typetree, dune-functions, dune-grid-glue, dune-pdelab, dune-corepy, dune-fufem and all the others?

**Comment by andlaus** [comment]
maybe you should explicitly state what your problem with the present PRs is: if you're a happy user of the current build system, nothing changes for you, if you have troubles you have a second option. what's so bad about having second options? I also do not expect anybody who is not interested in the dune BS to fix issues there.
